### PR TITLE
[Local App Snippet] support non conversational LLMs

### PR DESCRIPTION
**Pull request by mishig25**
## Description

Most GGUF files on the hub are insutrct/conversational. However, not all of them. Previously, local app snippets assumed that all GGUFs are insutrct/conversational.

### vLLM

https://huggingface.co/meta-llama/Llama-3.2-3B?local-app=vllm

```json
mishig@machine:~$ curl -X POST "http://localhost:8000/v1/completions" \
        -H "Content-Type: application/json" \
        --data '{
                "model": "meta-llama/Llama-3.2-3B",
                "prompt": "Once upon a time",
                "max_tokens": 150,
                "temperature": 0.5
        }'

{"id":"cmpl-157aad50ba6d45a5a7e2641a3c8157dd","object":"text_completion","created":1728293162,"model":"meta-llama/Llama-3.2-3B","choices":[{"index":0,"text":" there was a man who was very generous and kind to everyone. He was a good man and a good person. One day he was walking down the street and he saw a man who was very poor and starving. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking. The man was so hungry that he was crying and shaking","logprobs":null,"finish_reason":"length","stop_reason":null,"prompt_logprobs":null}],"usage":{"prompt_tokens":5,"total_tokens":155,"completion_tokens":150}}
```

### llama.cpp

https://huggingface.co/mlabonne/gemma-2b-GGUF?local-app=llama.cpp

```
llama-cli \
  --hf-repo "mlabonne/gemma-2b-GGUF" \
  --hf-file gemma-2b.Q2_K.gguf \
  -p "Once upon a time "
```

### llama-cpp-python

```python
from llama_cpp import Llama

llm = Llama.from_pretrained(
        repo_id="mlabonne/gemma-2b-GGUF",
        filename="gemma-2b.Q2_K.gguf",
)

output = llm(
        "Once upon a time ",
        max_tokens=512,
        echo=True
)

print(output)
```

### CODE REVIEW

File: packages/tasks/src/local-apps.ts [hunk 1]
```diff
@@ -83,15 +83,19 @@ function isLlamaCppGgufModel(model: ModelData) {
 }
 
 const snippetLlamacpp = (model: ModelData, filepath?: string): LocalAppSnippet[] => {
-	const command = (binary: string) =>
-		[
+	const command = (binary: string) => {
+		let snippet = [
```
Comment (julien-c) [inline packages/tasks/src/local-apps.ts:87]: can you keep it a const, by swithing the place of the `.join()` at the very end, i.e. pushing to the array inside the `if`?

File: packages/tasks/src/model-libraries-snippets.ts [hunk 2]
```diff
@@ -360,23 +360,36 @@ backbone = keras_nlp.models.Backbone.from_preset("hf://${model.id}")
 `,
 ];
 
-export const llama_cpp_python = (model: ModelData): string[] => [
-	`from llama_cpp import Llama
+export const llama_cpp_python = (model: ModelData): string[] => {
+	let snippet = `from llama_cpp import Llama
```
Comment (julien-c) [inline packages/tasks/src/model-libraries-snippets.ts:364]: switching to an array might be more legible

File: packages/tasks/src/local-apps.ts [hunk 3]
```diff
@@ -143,17 +147,29 @@ const snippetLocalAI = (model: ModelData, filepath?: string): LocalAppSnippet[]
...
 		`		]`,
 		`	}'`,
 	];
+	const runCommandNonInstruct = [
+		"# Call the server using curl:",
+		`curl -X POST "http://localhost:8000/v1/completions" \\`,
+		`	-H "Content-Type: application/json" \\`,
+		`	--data '{`,
+		`		"model": "${model.id}",`,
+		`		"prompt": "Once upon a time ",`,
```
Comment (Vaibhavs10) [inline packages/tasks/src/local-apps.ts:167]:
```suggestion
		`		"prompt": "Once upon a time",`,
```

Remove trailing space in the prompt, adding a space almost always results in weird generations.

File: packages/tasks/src/model-libraries-snippets.ts [hunk 4]
```diff
@@ -360,23 +360,36 @@ backbone = keras_nlp.models.Backbone.from_preset("hf://${model.id}")
...
 			"role": "user",
 			"content": "What is the capital of France?"
 		}
 	]
-)`,
-];
+)`;
+	} else {
+		snippet += `output = llm(
+	"Once upon a time ",
```
Comment (Vaibhavs10) [inline packages/tasks/src/model-libraries-snippets.ts:384]:
```suggestion
	"Once upon a time",
```

Same as above

File: packages/tasks/src/local-apps.ts [hunk 5]
```diff
@@ -83,15 +83,19 @@ function isLlamaCppGgufModel(model: ModelData) {
...
-		[
+	const command = (binary: string) => {
+		let snippet = [
 			"# Load and run the model:",
 			`${binary} \\`,
 			`  --hf-repo "${model.id}" \\`,
 			`  --hf-file ${filepath ?? "{{GGUF_FILE}}"} \\`,
-			'  -p "You are a helpful assistant" \\',
-			"  --conversation",
+			`  -p "${model.tags.includes("conversational") ? "You are a helpful assistant" : "Once upon a time "}"`,
```
Comment (Vaibhavs10) [inline packages/tasks/src/local-apps.ts:92]:
```suggestion
			`  -p "${model.tags.includes("conversational") ? "You are a helpful assistant" : "Once upon a time"}"`,
```

**Review by Vaibhavs10 — COMMENTED** [review]
Minor nit, but important specially wrt llama.cpp

File: packages/tasks/src/local-apps.ts [hunk 6]
```diff
@@ -143,17 +147,29 @@ const snippetLocalAI = (model: ModelData, filepath?: string): LocalAppSnippet[]
...
 		`		]`,
 		`	}'`,
 	];
+	const runCommandNonInstruct = [
+		"# Call the server using curl:",
+		`curl -X POST "http://localhost:8000/v1/completions" \\`,
+		`	-H "Content-Type: application/json" \\`,
+		`	--data '{`,
+		`		"model": "${model.id}",`,
+		`		"prompt": "Once upon a time",`,
```
Comment (gary149) [inline packages/tasks/src/local-apps.ts:167]:
```suggestion
		`		"prompt": "Once upon a time,"`,
```

File: packages/tasks/src/model-libraries-snippets.ts [hunk 7]
```diff
@@ -360,23 +360,36 @@ backbone = keras_nlp.models.Backbone.from_preset("hf://${model.id}")
...
 			"role": "user",
 			"content": "What is the capital of France?"
 		}
 	]
-)`,
-];
+)`;
+	} else {
+		snippet += `output = llm(
+	"Once upon a time",
```
Comment (gary149) [inline packages/tasks/src/model-libraries-snippets.ts:384]:
```suggestion
	"Once upon a time,",
```

File: packages/tasks/src/local-apps.ts (hunk 6 above)
Comment (gary149) [inline packages/tasks/src/local-apps.ts:167]:
```suggestion
		`		"prompt": "Once upon a time,",`,
```

File: packages/tasks/src/model-libraries-snippets.ts (hunk 2 above)
Comment (mishig25) [reply packages/tasks/src/model-libraries-snippets.ts]: handled in https://github.com/huggingface/huggingface.js/pull/954/commits/b9baabc2cfd5ff28deddebf5bd0e29efc95e9d97

**Comment by mishig25** [comment]
Added test cases as the examples are getting more complex and we can be sure not to break any existing examples

File: packages/tasks/src/local-apps.ts [hunk 8]
```diff
@@ -143,17 +148,29 @@ const snippetLocalAI = (model: ModelData, filepath?: string): LocalAppSnippet[]
...
+	const runCommandInstruct = [
 		"# Call the server using curl:",
-		`curl -X POST "http://localhost:8000/v1/chat/completions" \\ `,
-		`	-H "Content-Type: application/json" \\ `,
+		`curl -X POST "http://localhost:8000/v1/chat/completions" \\`,
+		`	-H "Content-Type: application/json" \\`,
 		`	--data '{`,
 		`		"model": "${model.id}",`,
 		`		"messages": [`,
 		`			{"role": "user", "content": "Hello!"}`,
```
Comment (Vaibhavs10) [inline packages/tasks/src/local-apps.ts:158]:
```suggestion
		`			{"role": "user", "content": "What is the capital of France?"}`,
```

Minor suggestion: `Hello!` looks a bit too terse. Perhaps we can unify the Instruct examples to be the same as llama-cpp-python and so on.

File: packages/tasks/package.json [hunk 9]
```diff
@@ -30,6 +30,7 @@
 		"watch": "npm-run-all --parallel watch:export watch:types",
 		"prepare": "pnpm run build",
 		"check": "tsc",
+		"test": "vitest run",
```
Comment (julien-c) [inline packages/tasks/package.json:33]: is this correct @coyotte508?

Comment (coyotte508) [reply packages/tasks/package.json:33]: yes!

**Comment by mishig25** [comment]
This PR is finally ready to be reviewed.

Besides the changes described in [the description](https://github.com/huggingface/huggingface.js/pull/954#issue-2569897349), vLLM snippet also supports now vision models:

https://github.com/huggingface/huggingface.js/blob/d6c5b5b597732253a58c58c2912479fc21d91b6c/packages/tasks/src/local-apps.spec.ts#L98-L121

@Vaibhavs10 @pcuenca @julien-c

File: packages/tasks/src/local-apps.ts (hunk 8 above)
Comment (mishig25) [reply packages/tasks/src/local-apps.ts]: handled in https://github.com/huggingface/huggingface.js/pull/954/commits/2e7c0808ca75a10222d4fc9bcdda1fff7660aa19

File: packages/tasks/src/local-apps.ts [hunk 10]
```diff
@@ -178,22 +186,33 @@ const snippetLocalAI = (model: ModelData, filepath?: string): LocalAppSnippet[]
...
-		`curl -X POST "http://localhost:8000/v1/chat/completions" \\`,
-		`	-H "Content-Type: application/json" \\`,
-		`	--data '{`,
-		`		"model": "${model.id}",`,
-		`		"messages": [`,
-		`			{"role": "user", "content": "Hello!"}`,
-		`		]`,
-		`	}'`,
-	];
+	const messages = getModelInputSnippet(model) as ChatCompletionInputMessage[];
```
Comment (pcuenca) [inline packages/tasks/src/local-apps.ts:189]: ah nice!

**Review by pcuenca — COMMENTED** [review]
Conceptually looks great!

**Review by Vaibhavs10 — COMMENTED** [review]
Niceee! All good wrt `llama.cpp` + `llama-cpp-python` + `vllm` snippets. Do we need to standardise TGI snippets too?

**Comment by mishig25** [comment]
> Do we need to standardise TGI snippets too?

Yes, but lets do it in subseq PR once this PR gets merged into main

**Review by Vaibhavs10 — APPROVED** [review]
Merci!

**Pull request merged**